File: elixir/lib/nucleus/models/columnar.ex
defmodule Nucleus.Models.Columnar do
  @moduledoc """
  Columnar analytics model — COLUMNAR_INSERT, COLUMNAR_COUNT, COLUMNAR_SUM/AVG/MIN/MAX.

  ## Example

      alias Nucleus.Models.Columnar

      Columnar.insert(client, "events", %{user_id: 42, action: "click", value: 1.5})
      {:ok, count} = Columnar.count(client, "events")
      {:ok, total} = Columnar.sum(client, "events", "value")
      {:ok, avg} = Columnar.avg(client, "events", "value")
  """

  @type client :: Nucleus.Client.t()

  @doc "Inserts a row into a columnar table."
  @spec insert(client(), String.t(), map()) :: {:ok, boolean()} | {:error, term()}
  def insert(client, table, values) when is_map(values) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.insert") do
      json = Jason.encode!(values)

      case Nucleus.Client.query(client, "SELECT COLUMNAR_INSERT($1, $2)", [table, json]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the row count for a columnar table."
  @spec count(client(), String.t()) :: {:ok, integer()} | {:error, term()}
  def count(client, table) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.count") do
      case Nucleus.Client.query(client, "SELECT COLUMNAR_COUNT($1)", [table]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the sum of a column."
  @spec sum(client(), String.t(), String.t()) :: {:ok, number()} | {:error, term()}
  def sum(client, table, column) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.sum") do
      case Nucleus.Client.query(client, "SELECT COLUMNAR_SUM($1, $2)", [table, column]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the average of a column."
  @spec avg(client(), String.t(), String.t()) :: {:ok, float()} | {:error, term()}
  def avg(client, table, column) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.avg") do
      case Nucleus.Client.query(client, "SELECT COLUMNAR_AVG($1, $2)", [table, column]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the minimum value of a column."
  @spec min(client(), String.t(), String.t()) :: {:ok, term()} | {:error, term()}
  def min(client, table, column) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.min") do
      case Nucleus.Client.query(client, "SELECT COLUMNAR_MIN($1, $2)", [table, column]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the maximum value of a column."
  @spec max(client(), String.t(), String.t()) :: {:ok, term()} | {:error, term()}
  def max(client, table, column) do
    with :ok <- Nucleus.Client.require_nucleus(client, "Columnar.max") do
      case Nucleus.Client.query(client, "SELECT COLUMNAR_MAX($1, $2)", [table, column]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end
end

File: rs/crates/neutron-nucleus/src/models/columnar.rs
//! Columnar analytics model — COLUMNAR_INSERT, COLUMNAR_COUNT, COLUMNAR_SUM,
//! COLUMNAR_AVG, COLUMNAR_MIN, COLUMNAR_MAX.

use serde_json;

use crate::error::NucleusError;
use crate::models::is_valid_identifier;
use crate::pool::NucleusPool;

/// Handle for columnar analytics operations.
pub struct ColumnarModel {
    pool: NucleusPool,
}

impl ColumnarModel {
    pub(crate) fn new(pool: NucleusPool) -> Self {
        Self { pool }
    }

    /// Insert a row into a columnar table. Values is a JSON object of column->value.
    pub async fn insert(
        &self,
        table: &str,
        values: &serde_json::Value,
    ) -> Result<bool, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let values_str =
            serde_json::to_string(values).map_err(|e| NucleusError::Serde(e.to_string()))?;
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_INSERT($1, $2)", &[&table, &values_str])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, bool>(0))
    }

    /// Return the number of rows in a columnar table.
    pub async fn count(&self, table: &str) -> Result<i64, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_COUNT($1)", &[&table])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, i64>(0))
    }

    /// Return the sum of a column.
    pub async fn sum(&self, table: &str, column: &str) -> Result<f64, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_SUM($1, $2)", &[&table, &column])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, f64>(0))
    }

    /// Return the average of a column.
    pub async fn avg(&self, table: &str, column: &str) -> Result<f64, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_AVG($1, $2)", &[&table, &column])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, f64>(0))
    }

    /// Return the minimum value of a column (as a string for type flexibility).
    pub async fn min(&self, table: &str, column: &str) -> Result<String, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_MIN($1, $2)::TEXT", &[&table, &column])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, String>(0))
    }

    /// Return the maximum value of a column (as a string for type flexibility).
    pub async fn max(&self, table: &str, column: &str) -> Result<String, NucleusError> {
        if !is_valid_identifier(table) {
            return Err(NucleusError::InvalidIdentifier(table.to_string()));
        }
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT COLUMNAR_MAX($1, $2)::TEXT", &[&table, &column])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, String>(0))
    }
}
